### PR TITLE
[GPU] Add subsequent reshapes optimization and dynamic paddings support for RoPE and PagedAttention

### PR DESCRIPTION
**Pull request by sshlyapn**
### Details:
 - Added subsequent reshapes optimization
 - Added dynamic paddings support for RoPE and PagedAttention

### Tickets:
 - [CVS-156124](https://jira.devtools.intel.com/browse/CVS-156124)

### CODE REVIEW

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/sdpa_opt.cl [hunk 1]
```diff
@@ -834,8 +835,11 @@ KERNEL(sdpa_opt)(
         const uint heads_dim = num_heads_dim;
 #endif
         #define KEY_SEQ_OFFSET subsequence_begins[gws_seq_indexes_correspondence[target_seq_dim]]
-        uint key_offset = KEY_SEQ_OFFSET * HEAD_SIZE * NUM_KV_HEADS + heads_dim * HEAD_SIZE + seq_len * HEAD_SIZE * NUM_KV_HEADS;
-        const uint key_pitch = HEAD_SIZE * NUM_KV_HEADS;
+        uint key_offset = INPUT1_PAD_BEFORE_FEATURE_NUM +
+                          KEY_SEQ_OFFSET * (HEAD_SIZE * NUM_KV_HEADS + INPUT1_PAD_BEFORE_FEATURE_NUM + INPUT1_PAD_AFTER_FEATURE_NUM) +
+                          heads_dim * HEAD_SIZE +
+                          seq_len * (HEAD_SIZE * NUM_KV_HEADS + INPUT1_PAD_BEFORE_FEATURE_NUM + INPUT1_PAD_AFTER_FEATURE_NUM);
+        const uint key_pitch = (HEAD_SIZE * NUM_KV_HEADS + INPUT1_PAD_BEFORE_FEATURE_NUM + INPUT1_PAD_AFTER_FEATURE_NUM);
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/sdpa_opt.cl:842]: nit: `key_pitch` can be used to compute `key_offset`

**Review by vladimir-paramuzov — APPROVED** [review]
Overall, LGTM. 
Please consider merging OptimizeSubsequentReshapes with ReshapeSequenceFusion later

File: src/plugins/intel_gpu/src/plugin/transformations/optimize_subsequent_reshapes.cpp [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+namespace intel_gpu {
+
+OptimizeSubsequentReshapes::OptimizeSubsequentReshapes() {
+    using namespace ov::pass::pattern;
+    using ov::pass::pattern::op::Or;
+
+    auto dynamic_batch_only = [](Output<Node> output) {
+        const auto& shape = output.get_partial_shape();
+
+        if (shape.size() <= 1)
```
Comment (itikhono) [inline src/plugins/intel_gpu/src/plugin/transformations/optimize_subsequent_reshapes.cpp:25]: minor: also it's better to check that rank is static

File: src/plugins/intel_gpu/src/plugin/transformations/optimize_subsequent_reshapes.hpp [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+#pragma once
+
+#include "openvino/pass/graph_rewrite.hpp"
+
+namespace ov {
+namespace intel_gpu {
+
+/**
+ * @brief This pass looks for specific patterns of subsequent reshapes that can be
```
Comment (itikhono) [inline src/plugins/intel_gpu/src/plugin/transformations/optimize_subsequent_reshapes.hpp:13]: could you describe a new covered case in this description? 
I believe this is `Reshape [ dynamic dim, n static dims] -> Reshape [dynamic dim, n static dims]`?
can we mention here that  OptimizeSubsequentReshapes can be combined with ReshapeSequenceFusion? mb add a ticket number if any

**Review by itikhono — APPROVED** [review]

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/rope_ref.cl [hunk 4]
```diff
@@ -28,14 +28,11 @@ KERNEL(rope_ref)(
...
     uint f = rf < HEAD_SIZE - ROTARY_NDIMS ? rf * 2 : 0;
 
-#ifdef ENABLE_SLICE
-    uint input_idx = GET_DATA_INDEX(SLICED_INPUT0, p, b, h * HEAD_SIZE, 0);
-
-    input_idx += SLICED_FROM_START * (p * INPUT0_FEATURE_NUM + b + 1)
-              + SLICED_FROM_END * (p * INPUT0_FEATURE_NUM + b);
-#else
     uint input_idx = INPUT0_GET_INDEX(p, b, h * HEAD_SIZE, 0);
+#ifdef ENABLE_SLICE
```
Comment (Lyamin-Roman) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/rope_ref.cl:32]: If the new method of calculating indices is correct, then the constant definitions from rope_kernel_base.cpp should be removed too

Comment (sshlyapn) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/rope_ref.cl:32]: Removed, thanks

**Review by Lyamin-Roman — APPROVED** [review]

**Pull request merged**